File: Python/Singleton/TheOnlyOne/theonlyone.py
# TODO: Other safer and more pythonic implementations.


class TheOnlyOne:
    __only_instance = None

    def __new__(cls, *args, **kwargs):
        if not cls.__only_instance:
            print('No instance exists yet, creating the only instance...')
            cls.__only_instance = super().__new__(cls)
        print('This is the only instance!')
        return cls.__only_instance


if __name__ == '__main__':
    the_only_one = TheOnlyOne()
    another_one = TheOnlyOne()
    print(the_only_one)
    print(another_one)

File: C++/Singleton/TheOnlyOne/the_only_one.cc
//
// Created by Eric Chen on 12/14/19.
//

#include <iostream>
#include "the_only_one.h"

std::unique_ptr<TheOnlyOne> TheOnlyOne::only_instance {nullptr};

TheOnlyOne* TheOnlyOne::get_instance() {
    if (!only_instance) {
        std::cout << "No instance exists yet, creating the only instance...\n";
        only_instance.reset(new TheOnlyOne());
    }
    std::cout << "This is the only instance!\n";
    return only_instance.get();
}
